netty/src/main/java/io/grpc/netty/NettyServerTransport.java

```diff
@@ -164,12 +173,39 @@ private void notifyTerminated(Throwable t) {
     }
   }
 
+  @Override
+  @Nullable
+  public TransportTracer.Stats getTransportStats() {
+    if (channel.eventLoop().inEventLoop()) {
+      // This is necessary, otherwise we will block forever on the promise in the else case
+      return transportTracer.getStats();
+    }
+    final Future<TransportTracer.Stats> future = channel.eventLoop().submit(
+        new Callable<TransportTracer.Stats>() {
+          @Override
+          public TransportTracer.Stats call() throws Exception {
+            return transportTracer.getStats();
+          }
+        });
+    try {
+      // Arbitrarily chosen timeout to avoid blocking forever
+      return future.get(1, TimeUnit.MINUTES);
```

Comment:
This is dangerous, as it must not be on _any_ event loop (not just this transport's); otherwise you can forever-blocking between two different event loops (consider the cases that lock ordering tries to solve). I don't see any users of this API so I don't know what's best, but we should either: return a future or have a callback be passed in.

Reply:
This is meant to be called from the thread handling the channelz request. I agree making this a Future would better communicate the potential to block. We also get the nice side effect of being able to kick off many reads in parallel.